### PR TITLE
[SPARK-18845][GraphX] PageRank has incorrect initialization value that leads to slow convergence

### PR DESCRIPTION
**Pull request by aray**
## What changes were proposed in this pull request?

Change the initial value in all PageRank implementations to be `1.0` instead of `resetProb` (default `0.15`) and use `outerJoinVertices` instead of `joinVertices` so that source vertices get updated in each iteration. 

This seems to have been introduced a long time ago in https://github.com/apache/spark/commit/15a564598fe63003652b1e24527c432080b5976c#diff-b2bf3f97dcd2f19d61c921836159cda9L90

With the exception of graphs with sinks (which currently give incorrect results see SPARK-18847) this gives faster convergence as the sum of ranks is already correct (sum of ranks should be number of vertices).

Convergence comparision benchmark for small graph: http://imgur.com/a/HkkZf
Code for benchmark: https://gist.github.com/aray/a7de1f3801a810f8b1fa00c271a1fefd

## How was this patch tested?

(corrected) existing unit tests and additional test that verifies against result of igraph and NetworkX on a loop with a source.

### CODE REVIEW

**Comment by aray** [comment]
Updated the above benchmark code with a log normal random graph on 10,000 vertices the difference is much more drastic.

(take the very bottom of the graph with a grain of salt as its in comparison to `g.pageRank(0.00001)`, actual error continues to drop)

**Comment by aray** [comment]
ping @srowen @dbtsai @rxin @ankurdave @jegonzal

**Comment by srowen** [comment]
I am not sure who if anyone would review graphx at this point, and I am not so familiar with the implementation here. If it converges to the same answer faster that's good. it might be nice to understand why this init is better, like any paper or similar implementaiton.

**Comment by rxin** [comment]
I just emailed @ankurdave and he is going to look at this tonight.

**Comment by aray** [comment]
**References**
[Pagerank paper](http://ilpubs.stanford.edu:8090/422/1/1999-66.pdf)
> We need to make an initial assignment of the ranks. This assignment can be made by one of several strategies. If it is going to iterate until convergence, in general the initial values will not affect final values, just the rate of convergence. But we can speed
up convergence by choosing a good initial assignment.

Since they are more focused on updating values for one evolving graph (the internet) they dont really talk about starting from scratch. But this does emphisize that there is no change to answers, just rate of convergence.

A more direct statement would be [Wikipedia](https://en.wikipedia.org/wiki/PageRank)
> PageRank is initialized to the same value for all pages. In the original form of PageRank, the sum of PageRank over all pages was the total number of pages on the web at that time, so each page in this example would have an initial value of 1.

Note that there are two variants of pagerank that differ by a constant multiple in outputs but are determined by the dampening factor, we use the version that sums to N (most other implementations use the other). More Wikipedia:
>The difference between them is that the PageRank values in the first formula sum to one, while in the second formula each PageRank is multiplied by N and the sum becomes N.

Essentialy starting with the correct sum is closer to the actual fixed point and thus gets you faster convergence.

The [NetworkX implementation](https://github.com/networkx/networkx/blob/master/networkx/algorithms/link_analysis/pagerank_alg.py#L122) uses the variant that sums to 1 hence their initialization values are all 1/N. 

igraph is unfortunately not comparable as they use a [more complex linear solver approach](https://github.com/igraph/igraph/blob/master/src/prpack/prpack_solver.cpp)

Additional credentials (if it matters): PhD Mathematics with dissertation in Graph Theory

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 1]
```diff
@@ -225,11 +225,11 @@ object PageRank extends Logging {
         ctx => ctx.sendToDst(ctx.srcAttr :* ctx.attr),
         (a : BV[Double], b : BV[Double]) => a :+ b, TripletFields.Src)
 
-      rankGraph = rankGraph.joinVertices(rankUpdates) {
-        (vid, oldRank, msgSum) =>
-          val popActivations: BV[Double] = msgSum :* (1.0 - resetProb)
+      rankGraph = rankGraph.outerJoinVertices(rankUpdates) {
```
Comment (srowen) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Is this a related but slightly separate fix?

File: graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala [hunk 2]
```diff
@@ -70,10 +70,10 @@ class PageRankSuite extends SparkFunSuite with LocalSparkContext {
       val resetProb = 0.15
       val errorTol = 1.0e-5
 
-      val staticRanks1 = starGraph.staticPageRank(numIter = 1, resetProb).vertices
-      val staticRanks2 = starGraph.staticPageRank(numIter = 2, resetProb).vertices.cache()
+      val staticRanks1 = starGraph.staticPageRank(numIter = 2, resetProb).vertices
```
Comment (srowen) [inline graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala]: A few extra iterations makes the test more robust?

**Review by srowen — COMMENTED** [review]
I mostly plead ignorance, but seems reasonable. Is the improvement mostly coming from a better magnitude of the initial weights then?

I'm just trying to figure out if the current implementation is also a different, decent idea, and just scaling resetProb is even better. No idea.

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala (hunk 1 above)
Comment (aray) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Its an intertwined bug in the implementation that was introduced at the same time (when moving away from Pregel in https://github.com/apache/spark/commit/15a564598fe63003652b1e24527c432080b5976c). The only vertices not included in the original `joinVertices` are source vertices (those with no incoming edges). Normally (in the absence of sinks) source vertices would have page rank equal to the reset probability. Since source vertices were not included in the join their rank was fixed at their initial value, which fortunately was the correct value. When we change the initial value of all vertices to 1 it exposes this error.

File: graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala (hunk 2 above)
Comment (aray) [inline graphx/src/test/scala/org/apache/spark/graphx/lib/PageRankSuite.scala]: Not really more robust since it has a sink and thus is still wrong pending SPARK-18847. But it is needed with the change to fully propagate the change in rank of source vertices in the first iteration as explained above.

**Comment by aray** [comment]
Yes the improvement is from the sum of magnitudes of initial values being closer to the (known) sum of the solution. Fiddling with resetProb controls a completely different thing. The current implementation has no advantage (excluding finding the incorrect solution to a star graph one iteration faster).

**Comment by ankurdave** [comment]
Thanks @aray for the explanation. I agree with @srowen - this looks reasonable to me. I'm going to merge it.

**Comment by ankurdave** [comment]
Merged into master.
